File: trabalhos_praticos/TP03/Q01/ListaSequencial.java
/*
Trabalho Prático 03

@catfmcastro
Catarina F. M. Castro

AEDs II
*/

// lista de registros simples com jogadores

import java.util.*;
import java.io.*;

class Jogador {
    // atributos
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;

    // construtor sem parametro
    public Jogador() {
        this.id = -1;
        this.nome = "";
        this.altura = -1;
        this.peso = -1;
        this.universidade = "";
        this.anoNascimento = -1;
        this.cidadeNascimento = "";
        this.estadoNascimento = "";
    }

    // construtor com parametros
    public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento,
            String cidadeNascimento, String estadoNascimento) {
        this.id = id;
        this.nome = nome;
        this.altura = altura;
        this.peso = peso;
        this.universidade = universidade;
        this.anoNascimento = anoNascimento;
        this.cidadeNascimento = cidadeNascimento;
        this.estadoNascimento = estadoNascimento;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getNome() {
        return nome;
    }

    public void setAltura(int altura) {
        this.altura = altura;
    }

    public int getAltura() {
        return altura;
    }

    public void setPeso(int peso) {
        this.peso = peso;
    }

    public int getPeso() {
        return peso;
    }

    public void setUniversidade(String universidade) {
        this.universidade = universidade;
    }

    public String getUniversidade() {
        return universidade;
    }

    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }

    public int getAnoNascimento() {
        return anoNascimento;
    }

    public void setCidadeNascimento(String cidadeNascimento) {
        this.cidadeNascimento = cidadeNascimento;
    }

    public String getCidadeNascimento() {
        return cidadeNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento) {
        this.estadoNascimento = estadoNascimento;
    }

    public String getEstadoNascimento() {
        return estadoNascimento;
    }

    public void clone(Jogador x) {
        this.id = x.id;
        this.nome = x.nome;
        this.altura = x.altura;
        this.peso = x.peso;
        this.universidade = x.universidade;
        this.anoNascimento = x.anoNascimento;
        this.cidadeNascimento = x.cidadeNascimento;
        this.estadoNascimento = x.estadoNascimento;
    }

    public void imprimir() {
        System.out.println("[" + id + " ## " + nome + " ## " + altura + " ## " + peso + " ## " + anoNascimento + " ## "
                + universidade + " ## " + cidadeNascimento + " ## " + estadoNascimento + "]");
    }

    // ler uma str e converte para um Jogador
    public void ler(String str) {
        // separa as strings nas virgulas
        String[] atributos = str.split(",", -1);

        // id (sempre presente)
        this.id = Integer.parseInt(atributos[0]);

        // nome (sempre presente)
        this.nome = atributos[1];

        // altura
        if (atributos[2] != "") {
            this.altura = Integer.parseInt(atributos[2]);
        } else {
            this.altura = -1;
        }

        // peso
        if (atributos[3] != "") {
            this.peso = Integer.parseInt(atributos[3]);
        } else {
            this.peso = -1;
        }

        // universidade
        if (atributos[4] != "") {
            this.universidade = atributos[4];
        } else {
            this.universidade = "nao informado";
        }

        // anoNascimento
        if (atributos[5] != "") {
            this.anoNascimento = Integer.parseInt(atributos[5]);
        } else {
            this.anoNascimento = -1;
        }

        // cidadeNascimento
        if (atributos[6] != "") {
            this.cidadeNascimento = atributos[6];
        } else {
            this.cidadeNascimento = "nao informado";
        }

        // estadoNascimento
        if (atributos[7] != "") {
            this.estadoNascimento = atributos[7];
        } else {
            this.estadoNascimento = "nao informado";
        }

    }

}

class Celula {
    public Jogador elemento;
    public Celula prox;

    public Celula() {
    }

    public Celula(Jogador elemento) {
        this.elemento = elemento;
    }
}

class Lista {
    private Celula primeiro;
    private Celula ultimo;

    public int tamanho() {
        int tamanho = 0;
        for (Celula i = primeiro; i != ultimo; i = i.prox, tamanho++)
            ;
        return tamanho;
    }

    public void inserirInicio(Jogador x) {
        Celula tmp = new Celula(x);
        tmp.prox = primeiro.prox;
        primeiro.prox = tmp;
        if (primeiro == ultimo) {
            ultimo = tmp;
        }
        tmp = null;
    }

    public void inserirFim(Jogador x) {
        ultimo.prox = new Celula(x);
        ultimo = ultimo.prox;
    }

    public void inserir(Jogador x, int pos) throws Exception {
        int tam = tamanho();

        if (pos < 0 || pos > tam) {
            throw new Exception("Erro ao inserir posicao (" + pos + " / tamanho = " + tam + ") invalida!");
        } else if (pos == 0) {
            inserirInicio(x);
        } else if (pos == tam) {
            inserirFim(x);
        } else {
            // caminha ate a posicao anterior a insercao
            Celula i = primeiro;
            for (int j = 0; j < pos; j++, i = i.prox)
                ;

            Celula tmp = new Celula();
            tmp.prox = i.prox;
            i.prox = tmp;
            tmp = i = null;
        }
    }

    public Jogador removerInicio() throws Exception {
        if (primeiro == ultimo) {
            throw new Exception("Erro ao remover: lista vazia!");
        }

        Celula tmp = primeiro;
        primeiro = primeiro.prox;
        Jogador removido = primeiro.elemento;
        tmp.prox = null;
        tmp = null;

        return removido;
    }

    public Jogador removerFim() throws Exception {
        if (primeiro == ultimo) {
            throw new Exception("Erro ao remover: lista vazia!");
        }

        // caminhar ate a penultima celula
        Celula i;
        for (i = primeiro; i.prox != ultimo; i = i.prox)
            ;

        Jogador removido = ultimo.elemento;
        ultimo = i;
        i = ultimo.prox = null;

        return removido;
    }

    public Jogador remover(int pos) throws Exception {
        Jogador removido;
        int tam = tamanho();

        if (primeiro == ultimo) {
            throw new Exception("Erro ao remover: lista vazia!");
        } else if (pos < 0 || pos > tam) {
            throw new Exception("Erro ao remover posicao (" + pos + " / tamanho = " + tam + ") invalida!");
        } else if (pos == 0) {
            removido = removerInicio();
        } else if (pos == tam) {
            removido = removerFim();
        } else {
            // caminhar ate a posicao anterior a remocao
            Celula i = primeiro;
            for (int j = 0; j < pos; j++, i = i.prox)
                ;

            Celula tmp = i.prox;
            removido = tmp.elemento;
            i.prox = tmp.prox;
            tmp.prox = null;
            i = tmp = null;
        }

        return removido;
    }

    // public String mostrar(int id) {
        
    // }

}

public class ListaSequencial {

    // ler csv e converte para string
    public static String[] lerCsv(String path) {
        // "/tmp/" +
        File file = new File(path);

        // array de strings, com tamanho especifico do arquivo
        String[] csvData = new String[3922];

        // contador
        int i = 0;

        // le o arquivo e salva cada linha no array de strings
        try {
            Scanner sc = new Scanner(file);

            while (sc.hasNext()) {
                csvData[i] = sc.nextLine();
                i++;
            }

            // fechamento camera
            sc.close();

        } catch (FileNotFoundException e) {
            System.out.println("Não foi possível econtrar o arquivo: " + e.getMessage());
        }

        return csvData;
    }

    // método swap
    public static void swap(int x, int y, Jogador[] array) {
        Jogador tmp = array[x];
        array[x] = array[y];
        array[y] = tmp;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        String[] csvData = lerCsv("playersAtualizado.csv");

        // array com todos os jogadores
        Jogador[] jogadores = new Jogador[csvData.length];
        for (int i = 0; i < csvData.length; i++) {
            String aux = csvData[i];
            Jogador player = new Jogador();
            player.ler(aux);
            jogadores[i] = player;
        }

        // lista de jogadores
        Lista lista = new Lista();
        
        // insere jogadores na lista
        while (!sc.hasNext("FIM")) {
            int input = sc.nextInt();
            int id = input;
            lista.inserirFim(jogadores[id]);
        }
        
             

        // imprime o resultado na tela
        // for (int i = 0; i < tam; i++) {
        //     data[i].imprimir();
        // }

        // fechamento do scanner
        sc.close();
    }
}

File: trabalhos_praticos/TP03/Q02/ListaSequencial.c
/*
Trabalho Prático 03

@catfmcastro
Catarina F. M. Castro

AEDs II
*/

// lista sequencial de registros de jogadores

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// JOGADOR ---------------------------------------------------------------------------------------------------------------------
typedef struct Jogador {
    int id;
    char nome[100];
    char altura[100];
    char peso[100];
    char universidade[100];
    char anoNascimento[100];
    char cidadeNascimento[100];
    char estadoNascimento[100];
} Jogador;

int tam = 0;

Jogador clone(Jogador *jogador) {
    Jogador aux = *jogador;
    return aux;
}

// imprime um jogador na tela
void imprimir(Jogador jogador, int pos) {
    // converte string em int
    int altura = atoi(jogador.altura);
    int peso = atoi(jogador.peso);
    int anoNascimento = atoi(jogador.anoNascimento);

    printf("[%d] ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n", pos,
        jogador.nome, altura, peso, anoNascimento, jogador.universidade,
        jogador.cidadeNascimento, jogador.estadoNascimento);
}

char *linhaCsv(char *frase) {
    char noInfo[] = "nao informado";
    char *newfrase = malloc(sizeof(char) * 200);

    if (newfrase == NULL) {
        return NULL;
    }

    strcpy(newfrase, ""); // inicialização

    for (int i = 0; frase[i] != '\0'; i++) {
        if (frase[i] == ',' && frase[i + 1] == ',') {
            strcat(newfrase, ",");
            strcat(newfrase, noInfo);
        } else if (frase[i] == ',' && frase[i + 2] == '\0') {
            strcat(newfrase, ",");
            strcat(newfrase, noInfo);
        } else {
            char temp[2] = {frase[i], '\0'};
            strcat(newfrase, temp);
        }
    }

    return newfrase;
}

// add novo jogador
void addJogador(Jogador *jogador, char tokens[8][100]) {
    jogador->id = atoi(tokens[0]);

    strcpy(jogador->nome, tokens[1]);
    strcpy(jogador->altura, tokens[2]);
    strcpy(jogador->peso, tokens[3]);
    strcpy(jogador->universidade, tokens[4]);
    strcpy(jogador->anoNascimento, tokens[5]);
    strcpy(jogador->cidadeNascimento, tokens[6]);
    strcpy(jogador->estadoNascimento, tokens[7]);
}

// metodo split
void split(const char *str, char flag, char tokens[8][100]) {
    int line = 0;
    int index = 0;

    while (line < 8) {
        int i = 0;
        while (1) {
            if (str[index] == flag || str[index] == '\0' || str[index] == '\n') {
                tokens[line][i] = '\0';
                break;
            }

            tokens[line][i] = str[index];
            i++;
            index++;
        }

        index++;
        line++;
    }
}

void swap(int *x, int *y) {
    int tmp = *x;
    *x = *y;
    *y = tmp;
}

// LISTA --------------------------------------------------------------------------------------------------------------------

void inserirInicio(Jogador J, Jogador array[]) {
    for (int i = tam; i > 0; i--) {
        array[i] = array[i - 1];
    }
    array[0] = J;
    tam++;
}

void inserirFinal(Jogador J, Jogador array[]) {
    array[tam] = J;
    tam++;
}

void inserir(int posicao, Jogador J, Jogador array[]) {
    for (int i = tam; i > posicao; i--) {
        array[i] = array[i - 1];
    }
    array[posicao] = J;
    tam++;
}

void removerInicio(Jogador array[]) {
    Jogador tmp = array[0];
    for (int i = 0; i < tam - 1; i++) {
        array[i] = array[i + 1];
    }
    tam--;
    printf("(R) %s\n", tmp.nome);
}

void removerFinal(Jogador array[]) {
    Jogador tmp = array[tam - 1];
    array[tam - 1] = (Jogador){0, '\0', 0, 0, '\0', 0, '\0', '\0'};
    tam--;
    printf("(R) %s\n", tmp.nome);
}

void remover(int pos, Jogador array[]) {
    Jogador tmp = array[pos];
    for (int i = pos; i < tam - 1; i++)
    {
        array[i] = array[i + 1];
    }
    array[tam - 1] = (Jogador){0, '\0', 0, 0, '\0', 0, '\0', '\0'};
    tam--;
    printf("(R) %s\n", tmp.nome);
}

// MAIN ---------------------------------------------------------------------------------------------------------------------
int main (void) {
    long start = clock();
    char csvData[600];
    Jogador jogadores[3922];

    // leitura de arquivo csv
    // /tmp/
    FILE *arq = fopen("/tmp/players.csv", "r");
    fgets(csvData, sizeof(csvData), arq);
    for (int i = 0; fgets(csvData, 600, arq) != NULL; i++) {
        char *frase = linhaCsv(csvData);

        char dados[8][100];
        split(frase, ',', dados);
        free(frase);

        addJogador(&jogadores[i], dados);
    }

    Jogador *lista;
    lista = malloc(sizeof(Jogador) * 1000);
    int size = 0;
    
    // input de jogadores
    for (int i = 0; 1; i++) {
        size = i;
        char input[100];
        scanf("%s", input);
        if (strcmp(input, "FIM") == 0) {
            break;
        } else {
            int id = atoi(input);
            tam++;
            lista[i] = clone(&jogadores[id]);
        }
    }

    // comandos
    int n;
    scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        char comando[10];
        scanf("%s", comando);

        if (strcmp(comando, "II") == 0) {
            int id;
            scanf("%d", &id);
            inserirInicio(jogadores[id], lista);
        } else if (strcmp(comando, "IF") == 0) {
            int id;
            scanf("%d", &id);
            inserirFinal(jogadores[id], lista);
        } else if (strcmp(comando, "I*") == 0) {
            int pos, id;
            scanf("%d %d", &pos, &id);
            inserir(pos, jogadores[id], lista);
        } else if (strcmp(comando, "RI") == 0) {
            removerInicio(lista);
        } else if (strcmp(comando, "RF") == 0) {
            removerFinal(lista);
        }
        else if (strcmp(comando, "R*") == 0) {
            int pos;
            scanf("%d", &pos);
            remover(pos, lista);
        }
    }

    // saida
    for (int i = 0; i < tam; i++) {
        imprimir(lista[i], i);
    }

    // fechamento arquivo
    fclose(arq);

    return 0;
}